File: crawler/src/providers/mingyan_provider.ts
import { load } from "../engine/parser";
import { fetchWithDelay } from "../engine/http_client";
import { computeSimHash, computeFingerprint } from "../engine/deduper";
import BaseProvider, { RawQuote, EnrichedQuote } from "./base_provider";
import logger from "../logger";

const CATEGORIES: Record<string, string> = {
  life: "/shenghuo/",
  love: "/aiqing/",
  work: "/zhichang/",
  study: "/xuexi/",
  friendship: "/youqing/",
  family: "/qinqing/",
  success: "/chenggong/",
  time: "/shijian/",
};

class MingyanProvider extends BaseProvider {
  readonly code = "mingyan";
  readonly name = "名言通";
  readonly baseUrl = "https://www.mingyantong.com";

  async fetchList(page: number): Promise<RawQuote[]> {
    const items: RawQuote[] = [];

    for (const [categoryKey, path] of Object.entries(CATEGORIES)) {
      const url = `${this.baseUrl}${path}list-${page}.html`;
      logger.info(`[${this.code}] 抓取列表: ${url}`);

      try {
        const resp = await fetchWithDelay(url);
        const $ = load(resp.body);

        $(".quote-item, .list-item, .mingyu-list li").each((_, el) => {
          const $el = $(el);
          const content =
            $el.find(".quote-content, .content, p").first().text().trim();
          const author =
            $el.find(".quote-author, .author, .source").first().text().trim() || undefined;
          const href = $el.find("a").first().attr("href") || "";

          if (content && content.length > 5) {
            items.push({
              content,
              author,
              source: this.name,
              url: href.startsWith("http") ? href : `${this.baseUrl}${href}`,
              categoryType: categoryKey,
              tags: [categoryKey],
            });
          }
        });

        if (items.length === 0) {
          $("article, .quote, .item").each((_, el) => {
            const $el = $(el);
            const content = $el.text().trim();
            if (content && content.length > 10 && content.length < 500) {
              items.push({
                content,
                source: this.name,
                url: `${this.baseUrl}${path}`,
                categoryType: categoryKey,
                tags: [categoryKey],
              });
            }
          });
        }
      } catch (err: any) {
        logger.error(`[${this.code}] 列表抓取失败 ${url}: ${err.message}`);
      }
    }

    return items;
  }

  async fetchDetail(raw: RawQuote): Promise<EnrichedQuote> {
    const simHash = computeSimHash(raw.content);
    const fingerprint = computeFingerprint(raw.content);

    return {
      ...raw,
      simHash,
      fingerprint,
      wordCount: raw.content.length,
    };
  }
}

export default MingyanProvider;

File: crawler/crawler/providers/mingyan_provider.py
"""名言通 Provider

对应 TypeScript 版 crawler/src/providers/mingyan_provider.ts。
"""
from typing import List

from ..engine.deduper import compute_fingerprint, compute_sim_hash
from ..engine.http_client import fetch_with_delay
from ..engine.parser import load
from ..logger import logger
from .base_provider import BaseProvider, EnrichedQuote, RawQuote

CATEGORIES = {
    "life": "/shenghuo/",
    "love": "/aiqing/",
    "work": "/zhichang/",
    "study": "/xuexi/",
    "friendship": "/youqing/",
    "family": "/qinqing/",
    "success": "/chenggong/",
    "time": "/shijian/",
}


def _first_text(el, selector: str) -> str:
    """等价 TS $el.find(selector).first().text().trim()"""
    found = el.select(selector)
    if not found:
        return ""
    return found[0].get_text().strip()


class MingyanProvider(BaseProvider):
    """对应 TS MingyanProvider"""

    code = "mingyan"
    name = "名言通"
    baseUrl = "https://www.mingyantong.com"

    def fetch_list(self, page: int) -> List[RawQuote]:
        items: List[RawQuote] = []

        for category_key, path in CATEGORIES.items():
            url = f"{self.baseUrl}{path}list-{page}.html"
            logger.info(f"[{self.code}] 抓取列表: {url}")

            try:
                resp = fetch_with_delay(url)
                soup = load(resp.body)

                # 对应 TS $(".quote-item, .list-item, .mingyu-list li").each(...)
                for el in soup.select(".quote-item, .list-item, .mingyu-list li"):
                    content = _first_text(el, ".quote-content, .content, p")
                    author = _first_text(el, ".quote-author, .author, .source")
                    a = el.find("a")
                    href = (a.get("href") or "") if a is not None else ""

                    if content and len(content) > 5:
                        item: RawQuote = {
                            "content": content,
                            "source": self.name,
                            "url": href if href.startswith("http") else f"{self.baseUrl}{href}",
                            "categoryType": category_key,
                            "tags": [category_key],
                        }
                        if author:
                            item["author"] = author
                        items.append(item)

                if not items:
                    # 对应 TS 兜底 $("article, .quote, .item")
                    for el in soup.select("article, .quote, .item"):
                        content = el.get_text().strip()
                        if content and 10 < len(content) < 500:
                            items.append({
                                "content": content,
                                "source": self.name,
                                "url": f"{self.baseUrl}{path}",
                                "categoryType": category_key,
                                "tags": [category_key],
                            })
            except Exception as err:
                logger.error(f"[{self.code}] 列表抓取失败 {url}: {err}")

        return items

    def fetch_detail(self, raw: RawQuote) -> EnrichedQuote:
        content = raw.get("content") or ""
        sim_hash = compute_sim_hash(content)
        fingerprint = compute_fingerprint(content)

        enriched: EnrichedQuote = dict(raw)  # type: ignore[assignment]
        enriched["simHash"] = sim_hash
        enriched["fingerprint"] = fingerprint
        enriched["wordCount"] = len(content)
        return enriched
